server/src/main/java/org/opensearch/gateway/remote/model/RemoteObjectStore.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote.model;
+
+import org.opensearch.core.action.ActionListener;
+
+import java.io.IOException;
+
+/**
+ * An interface to read/write an object from/to a remote storage. This interface is agnostic of the remote storage type.
+ *
+ * @param <T> The object type which can be uploaded to or downloaded from remote storage.
+ */
```

Comment:
The comment seems inconsistent, aren't you uploading `U`?

Reply:
U is a wrapper object for T. We are actually uploading T to remote store. I can improve the comment.

Reply:
```
  public void writeAsync(U obj, ActionListener<Void> listener);
```

The interface is unclear, lets see a concrete implementation of this class

Reply:
@soosinha why do you need wrapper as U?

Reply:
@shwetathareja Did not understand the question completely.
U is an implementation of RemoteObject class. For example with T as IndexMetadata, U will be RemoteIndexMetadata.
For uploading, RemoteIndexMetadata will be initialized using IndexMetadata as a constructor param and then RemoteIndexMetadata will `serialize` IndexMetadata before uploading.
For downloading, RemoteIndexMetadata will be initialized using the blob name as a constructor param and then RemoteIndexMetadata will `deserialize` IndexMetadata after downloading.

Reply:
I think what you need is the below interface. Please remove `U` it is just a holder for the blob name/path
```
public interface RemoteObjectStore<T extends RemoteObject<T>> {

    public void writeAsync(T obj, ActionListener<Void> listener);

    public T read(Tuple<String, String> blobPath) throws IOException;

    public void readAsync(Tuple<String, String> blobPath, ActionListener<T> listener);
}

```

Reply:
```
public interface RemoteObjectStore<T extends RemoteObject<T>> {
```
In the above interface, the first T and second T are not the same.
When I create and instance of the store, it will look like below:
```
RemoteBlobStore<RemoteIndexMetadata> indexMetadataBlobStore = new RemoteBlobStore<>(blobStoreTransferService, blobStoreRepository, clusterName, threadPool);
```
Here the first T is RemoteIndexMetadata and second T is IndexMetadata.